daml-lf/language/src/main/scala/com/digitalasset/daml/lf/language/Ast.scala

```diff
@@ -1031,26 +994,63 @@ object Ast {
   type TemplateImplementsSignature = GenTemplateImplements[Unit]
   val TemplateImplementsSignature = new GenTemplateImplementsCompanion[Unit]
 
-  final case class GenTemplateImplementsMethod[E](
+  final case class GenInterfaceInstanceBody[E](
```

Comment:
Not a criticism of this PR... but a general comment on the `Ast.scala`...
I really hate this parameterization on`[E]` 

Where do we reap any benefit from this complication?

Reply:
the only `E` other than `Expr` that I'm aware of is `Unit`, which is used to define "signature" versions of each type, which is what `PackageInterface`'s lookups return. I haven't run into such an error myself but it might be useful to check an `x[Expr]` against an `x[Unit]` signature without mixing them up, perhaps a minor benefit. 

on the other hand, from a quick look at the codebase it seems we always get `xSignature` from `x` using the `toSignature` functions in `daml-lf/language/src/main/scala/com/digitalasset/daml/lf/language/Util.scala`, so I'm not sure how bad the penalty is for that benefit.

Reply:
We had a few issues with very high memory usage when storing all values in a package. That’s why we switched to being able to erase that and only keep compiled packages + the signature.

Reply:
thanks for the context @cocreature!